### PR TITLE
SAK-51153 Sitestats: Event collection duplicates event processing

### PR DESCRIPTION
**Pull request by st-manu**
https://sakaiproject.atlassian.net/browse/SAK-51153

We found this while testing. This is a real case:

---

**17:21:48.463** → Finished reading events **3137257703 to 3137268571**. A total of **10,868** events were read (`sqlBlockSize` is **10,000**) (**868 events have no session**).  

**17:23:56.049** → Finished processing events up to **3137268571**.  

**17:23:56.477** → Reads event ID **3137278134** for the first time. The event is at position **10,431**, exceeding the queue.  
**17:23:56.495** → Finished reading events **3137267703 to 3137278651**. A total of **10,948** events were read (**948 events have no session**).  
(The last event processed in the previous queue was **3137268571**, but now it starts at **3137267703**, repeating **868** events.)

**17:26:14.342** → Finished processing events up to **3137278651**.  

**17:26:14.361** → Reads event ID **3137278134** for the second time. The event is at position **431**.  
**17:26:14.750** → Finished reading events **3137277703 to 3137288542**. A total of **10,839** events were read (**839 events have no session**).  
(The last event processed in the previous queue was **3137278651**, but now it starts at **3137277703**, repeating **948** events.)

**17:28:23.217** → Finished processing events up to **3137288542**.

---

The issue is that the SQL query returns exactly `sqlBlockSize` rows, excluding events from invalid sessions. As a result, if `sqlBlockSize` is 1000 and the first `eventId` is 5000, and the query encounters 100 invalid sessions, it will return events from 5000 to 6100. However, the next batch starts collecting from 6000 (5000 + 1000), causing 100 events to be processed twice.

I've added an `if` statement that should never be reached, but it doesn't hurt to make sure.

### CODE REVIEW

File: sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/StatsAggregateJobImpl.java [hunk 1]
```diff
@@ -346,8 +360,10 @@ private String startJob() throws SQLException {
 				if(processedCounter >= getMaxEventsPerRun()) {
 					break;
 				} else {
-					offset += sqlBlockSize;
+					offset = lastProcessedEventIdWithSuccess + 1;
```
Comment (ottenhoff) [inline sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/StatsAggregateJobImpl.java:363]: This patch looks okay to me, but this line worries me. I remember many years ago we ran into an issue where the sitestats would just process the same events. Maybe there was an error in the processing and it would just process the same events without moving on. Is this line at least incrementing the counter 1 no matter what?

Comment (st-manu) [reply sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/StatsAggregateJobImpl.java:363]: yes, I've reviewed the jiras and seen the issue. I was also concerned that it might enter a loop. I did a quick test to see what happens if all the events in the queue have no valid session, and it increments correctly. I also tested the case where none of the events to be processed (not just in one queue) have valid sessions, all tests were successful

**Review by ottenhoff — APPROVED** [review]
